### PR TITLE
SONARJAVA-2494 Implement JVM instructions

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 1]
```diff
@@ -157,79 +133,154 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+        break;
+      case DUP2:
+        pop = programState.unstackValue(2);
+        Preconditions.checkState(pop.values.size() == 2, "DUP2 needs 2 values on stack");
+        programState = pop.state.stackValue(pop.values.get(1)).stackValue(pop.values.get(0)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
+        break;
+      case DUP2_X1:
+        pop = programState.unstackValue(3);
+        Preconditions.checkState(pop.values.size() == 3, "DUP2_X1 needs 3 values on stack");
+        programState = pop.state.stackValue(pop.values.get(1)).stackValue(pop.values.get(0)).stackValue(pop.values.get(2)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MINOR Split this 183 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 2]
```diff
@@ -157,79 +133,154 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+        break;
+      case DUP2_X2:
+        pop = programState.unstackValue(4);
+        Preconditions.checkState(pop.values.size() == 4, "DUP2_X2 needs 4 values on stack");
+        programState = pop.state.stackValue(pop.values.get(1)).stackValue(pop.values.get(0)).stackValue(pop.values.get(3)).stackValue(pop.values.get(2)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MINOR Split this 213 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L131](https://github.com/SonarSource/sonar-java/blob/0462eb9afa0cffb52d27617357f2638a4dabe07c/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L131): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 3]
```diff
@@ -24,6 +24,7 @@
 import com.google.common.collect.Lists;
 import org.objectweb.asm.Opcodes;
 import org.objectweb.asm.Type;
+import org.objectweb.asm.util.Printer;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MINOR Remove this unused import 'org.objectweb.asm.util.Printer'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/c8a46a7e2b7d0403d59db5877df2f8c847b8cba8/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/b9b87512847fc117c6698fe819f975f402ee4bec/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L131](https://github.com/SonarSource/sonar-java/blob/cda5b500a833898ea10770c340fc4d58ef6e8a85/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L131): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/f1dadf1f165bdaa1b42077685b1b9d681c52894e/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 4]
```diff
@@ -157,79 +134,269 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+      case DRETURN:
+      case ARETURN:
+        programState.storeExitValue();
+        programState = programState.unstackValue(1).state;
+        break;
+      case RETURN:
+        // do nothing
+        break;
+      case GETSTATIC:
+        // TODO SONARJAVA-2510 associated symbolic value with symbol
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 5]
```diff
@@ -157,79 +134,269 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
       return;
     }
     ProgramState.Pop pop;
+    SymbolicValue sv;
     switch (instruction.opcode) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Reduce the number of non-empty switch cases from 35 to at most 30. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1479)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/47ccb68a4b0c4cfb8e8ed001a1c531b1fab38b3a/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/3c4021745dacd4099213335db6d0f1fa2797edbc/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/6c4588a6c8b87770e4ff76850117897f25863278/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 6]
```diff
@@ -259,21 +418,81 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+        pop = programState.unstackValue(1);
+        Preconditions.checkState(pop.values.size() == 1, "INSTANCEOF needs 1 value on stack");
+        SymbolicValue.InstanceOfSymbolicValue instanceOf = new SymbolicValue.InstanceOfSymbolicValue();
+        instanceOf.computedFrom(pop.valuesAndSymbols);
+        programState = pop.state.stackValue(instanceOf);
+        break;
+      case MONITORENTER:
+      case MONITOREXIT:
+        pop = programState.unstackValue(1);
+        Preconditions.checkState(pop.values.size() == 1, Printer.OPCODES[instruction.opcode] + " needs 1 value on stack");
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Use the built-in formatting to construct this argument. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 7]
```diff
@@ -259,21 +418,81 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+        programState = pop.state;
+        break;
+      case MULTIANEWARRAY:
+        BytecodeCFGBuilder.MultiANewArrayInsn multiANewArrayInsn = (BytecodeCFGBuilder.MultiANewArrayInsn) instruction;
+        pop = programState.unstackValue(multiANewArrayInsn.dim);
+        Preconditions.checkState(pop.values.size() == multiANewArrayInsn.dim, "MULTIANEWARRAY needs " + multiANewArrayInsn.dim + " values on stack");
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Use the built-in formatting to construct this argument. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [BytecodeEGWalker.java#L132](https://github.com/SonarSource/sonar-java/blob/e3d2a28b5ef59729419849ffc6b9faa5906571a0/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L132): Refactor this method to reduce its Cognitive Complexity from 25 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 8]
```diff
@@ -152,130 +130,380 @@ private void execute(String signature, SquidClassLoader classLoader) {
   }
 
   @VisibleForTesting
-  void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
+  void executeInstruction(Instruction instruction) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 9]
```diff
@@ -152,130 +130,380 @@ private void execute(String signature, SquidClassLoader classLoader) {
...
+      case MONITORENTER:
+      case MONITOREXIT:
+        pop = programState.unstackValue(1);
+        Preconditions.checkState(pop.values.size() == 1, Printer.OPCODES[instruction.opcode] + " needs 1 value on stack");
+        programState = pop.state;
+        break;
+      case MULTIANEWARRAY:
+        Instruction.MultiANewArrayInsn multiANewArrayInsn = (Instruction.MultiANewArrayInsn) instruction;
+        pop = programState.unstackValue(multiANewArrayInsn.dim);
+        Preconditions.checkState(pop.values.size() == multiANewArrayInsn.dim, "MULTIANEWARRAY needs " + multiANewArrayInsn.dim + " values on stack");
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Use the built-in formatting to construct this argument. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 10]
```diff
@@ -152,130 +130,380 @@ private void execute(String signature, SquidClassLoader classLoader) {
...
+      case INVOKESPECIAL:
+      case INVOKESTATIC:
+      case INVOKEINTERFACE:
+        if (handleMethodInvocation(instruction)) {
+          // when yields are available, do not execute post check on this node
+          return;
         }
         break;
+      case INVOKEDYNAMIC:
+        // TODO SONARJAVA-2512
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 11]
```diff
@@ -271,6 +272,63 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+        Preconditions.checkState(pop.values.size() == 1, "NEG needs value on stack");
+        programState = pop.state.stackValue(sv).addConstraint(sv, ObjectConstraint.NOT_NULL);
+        break;
+      case ISHL:
+      case LSHL:
+      case ISHR:
+      case LSHR:
+      case IUSHR:
+      case LUSHR:
+        handleArithmetic(instruction);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: As the implementation is the same for those, the cases should be next to the others.

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 12]
```diff
@@ -329,6 +329,33 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
...
+      case F2I:
+      case F2L:
+      case F2D:
+      case D2I:
+      case D2L:
+      case D2F:
+      case I2B:
+      case I2C:
+      case I2S:
+        // do nothing
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: Those instruction may have an impact on stack/values regarding : https://jira.sonarsource.com/browse/SONARJAVA-2508 
Maybe something should be added regarding those in that ticket ?

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 13]
```diff
@@ -356,12 +364,9 @@ void executeInstruction(BytecodeCFGBuilder.Instruction instruction) {
         Preconditions.checkState(pop.values.size() == 2, "CMP needs 2 values on stack");
         programState = pop.state.stackValue(sv).addConstraint(sv, ObjectConstraint.NOT_NULL);
         break;
-      case NEW: {
-        SymbolicValue symbolicValue = constraintManager.createSymbolicValue(instruction);
-        programState = programState.stackValue(symbolicValue);
-        programState = programState.addConstraint(symbolicValue, ObjectConstraint.NOT_NULL);
+      case NEW:
+        createNonNullValue(instruction);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: Merge case blocks with case LDC ?

File: java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java [hunk 14]
```diff
@@ -92,7 +92,8 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
   }
 
   public static class BytecodeCFG {
-    List<Block> blocks;
+    @VisibleForTesting
+    public List<Block> blocks;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java]: I am not sure I understand why this modification was actually required ?

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java (hunk 6 above)
Comment (benzonico) [reply java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: should be fixed.

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java (hunk 9 above)
Comment (benzonico) [reply java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: should be fixed

File: java-frontend/src/main/java/org/sonar/java/bytecode/cfg/Instruction.java [hunk 15]
```diff
@@ -81,6 +82,31 @@ public boolean hasReturnValue() {
     return Type.getMethodType(fieldOrMethod.desc).getReturnType() == Type.VOID_TYPE;
   }
 
+  @Override
+  public boolean equals(Object o) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/cfg/Instruction.java]: This is a bit of "going the extra mile" but I think it's worth it : I would add a test for those implementation of equals/hashcode.

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 16]
```diff
@@ -479,6 +448,45 @@ void executeInstruction(Instruction instruction) {
...
+          .forEach(yield ->
+            yield.statesAfterInvocation(
+              Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> constraintManager.createExceptionalSymbolicValue(yield.exceptionType())).forEach(ps -> {
+              ps.storeExitValue();
+              enqueue(new ProgramPoint(exitBlock), ps);
+            }));
+        return true;
+      }
+    }
+    if (instruction.hasReturnValue()) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: matter of taste and style but I tend to prefer : 
```
programState = pop.state;
if(!instruction.hasReturnValue()) {
   programState = programeState.stackValue(returnSV); 
}
```
one less branch...

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 17]
```diff
@@ -152,130 +130,380 @@ private void execute(String signature, SquidClassLoader classLoader) {
...
+        programState = pop.state.stackValue(pop.values.get(0)).stackValue(pop.values.get(2)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
+        break;
+      case DUP2:
+        pop = programState.unstackValue(2);
+        Preconditions.checkState(pop.values.size() == 2, "DUP2 needs 2 values on stack");
+        programState = pop.state.stackValue(pop.values.get(1)).stackValue(pop.values.get(0)).stackValue(pop.values.get(1)).stackValue(pop.values.get(0));
+        break;
+      case DUP2_X1:
+        pop = programState.unstackValue(3);
+        Preconditions.checkState(pop.values.size() == 3, "DUP2_X1 needs 3 values on stack");
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: I have the feeling that the part with 
```
  pop = programState.unstackValue(3);
  Preconditions.checkState(pop.values.size() == 3, "DUP2_X1 needs 3 values on stack");
``` 
should probably be extracted into a method doing the assertion with a generated message to simplify a bit the switch case code.

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java (hunk 12 above)
Comment (saberduck) [reply java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: yes, added TODO here, and updated ticket

File: java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java (hunk 14 above)
Comment (saberduck) [reply java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java]: leftover of some intermediate change, seems no longer necessary, reverted

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java (hunk 16 above)
Comment (saberduck) [reply java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: and the naming of the method was actually inversed

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 18]
```diff
@@ -152,130 +130,358 @@ private void execute(String signature, SquidClassLoader classLoader) {
...
+      case F2I:
+      case F2L:
+      case F2D:
+      case D2I:
+      case D2L:
+      case D2F:
+      case I2B:
+      case I2C:
+      case I2S:
+        // TODO SONARJAVA-2508 handle value conversion
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 19]
```diff
@@ -152,130 +130,358 @@ private void execute(String signature, SquidClassLoader classLoader) {
...
+          return;
         }
         break;
+      case INVOKEDYNAMIC:
+        // TODO SONARJAVA-2512
+        break;
+      case NEW:
+        createNonNullValue(instruction);
+        break;
+      case NEWARRAY:
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR This case's code block is the same as the block for the case on line 300. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**